test/taoensso/carmine/tests/message_queue.clj

```diff
@@ -118,29 +118,99 @@
   (is (= ["mid1" :msg1 1] (wcar* (dequeue* tq)))))
 
 (deftest test-7 ; Handling: enqueue while :done-with-backoff
-  (is (= "mid1" (do (clear-tq) (wcar* (mq/enqueue tq :msg1 :mid1)))))
+  (is (= "mid1" (do (clear-tq) (wcar* (mq/enqueue tq :msg1 :unique-message-id :mid1)))))
   (is (= "eoq-backoff" (wcar* (dequeue* tq))))
   (is (= :done-with-backoff
         (do (mq/handle1 conn-opts tq
               (fn [_] {:status :success :backoff-ms 3000})
               (wcar* (dequeue* tq)))
             (Thread/sleep 20)
             (wcar* (mq/message-status tq :mid1)))))
-  (is (= {:carmine.mq/error :done-with-backoff} (wcar* (mq/enqueue tq :msg1 :mid1))))
-  (is (= "mid1" (wcar* (mq/enqueue tq :msg1-requeued :mid1 :allow-requeue))))
+  (is (= {:carmine.mq/error :done-with-backoff} (wcar* (mq/enqueue tq :msg1 :unique-message-id :mid1))))
+  (is (= "mid1" (wcar* (mq/enqueue tq :msg1-requeued :unique-message-id :mid1 :allow-requeue? true))))
   (is (= :queued ; cmp :done-awaiting-gc
     (do (Thread/sleep 3000) ; Wait for backoff to expire
         (wcar* (mq/message-status tq :mid1)))))
   (is (= "eoq-backoff"    (wcar* (dequeue* tq))))
   (is (= ["mid1" :msg1 1] (wcar* (dequeue* tq)))))
+
+(deftest test-8 ; Enqueue/dequeue with initial backoff
+  (is (= "eoq-backoff" (do (clear-tq) (wcar* (dequeue* tq)))))
+  (is (= "mid1"        (wcar* (mq/enqueue tq :msg1 :unique-message-id :mid1 :initial-backoff-ms 500))))
+  (is (= "mid2"        (wcar* (mq/enqueue tq :msg2 :unique-message-id :mid2 :initial-backoff-ms 100))))
+  (is
+    (let [status (tq-status)
+          {:keys [messages mid-circle]} status]
+      (and
+        (= messages {"mid1" :msg1 "mid2" :msg2})
+        (= mid-circle ["mid2" "mid1" "end-of-circle"]))))
+  ;; Dupes before the backoff expired
+  (is (= {:carmine.mq/error :queued-with-backoff} (wcar* (mq/enqueue tq :msg1 :unique-message-id :mid1))))
+  (is (= {:carmine.mq/error :queued-with-backoff} (wcar* (mq/enqueue tq :msg2 :unique-message-id :mid2))))
+  ;; Both should be queued with backoff before the backoff expires
+  (is (= :queued-with-backoff        (wcar* (mq/message-status tq :mid1))))
+  (is (= :queued-with-backoff        (wcar* (mq/message-status tq :mid2))))
+  ;; Move time past second message
+  (is (do (Thread/sleep 150) true))
+  (is (= :queued-with-backoff        (wcar* (mq/message-status tq :mid1))))
+  (is (= :queued                     (wcar* (mq/message-status tq :mid2))))
+  ;; Move time past first message
+  (is (do (Thread/sleep 750) true))
+  (is (= :queued                     (wcar* (mq/message-status tq :mid1))))
+  (is (= :queued                     (wcar* (mq/message-status tq :mid2))))
+  ;; Dupes after the backoff expired
+  (is (= {:carmine.mq/error :queued} (wcar* (mq/enqueue tq :msg1 :unique-message-id :mid1))))
+  (is (= {:carmine.mq/error :queued} (wcar* (mq/enqueue tq :msg2 :unique-message-id :mid2))))
+  ;; TODO Is the order of retrieval actually predictable?
```

Comment:
Haven't looked at this closely, but from my recollection and quick inspection: **yes**, the order of retrieval should be predictable 👍 

The `mid-circle` used to select the next message-id for processing is a list (so ordered data structure).